File: Python/Dynamic Programming/L_predictTheWinner.py
from typing import List


def PredictTheWinner(nums: List[int]) -> bool:
        dp = list(nums)
        
        for length in range(2, len(nums)+1):
            for i in range(len(nums)-length+1):
                dp[i] =  max(nums[i]-dp[i+1], nums[i+length-1]-dp[i]);
          
        return dp[0] >= 0;

File: cpp/Dynamic Programming/L_predictTheWinner.cpp
#include <bits/stdc++.h>
using namespace std;

bool PredictTheWinner(vector<int> &nums)
{
    vector<int> dp(nums);

    for (int len = 2; len <= nums.size(); len++)
    {
        for (int i = 0; i + len - 1 < nums.size(); i++)
        {
            dp[i] = max(nums[i] - dp[i + 1], nums[i + len - 1] - dp[i]);
        }
    }

    return dp[0] >= 0;
}